src/libsystemd-network/ndisc-router.c

```diff
@@ -821,6 +824,124 @@ int sd_ndisc_router_captive_portal_get_uri(sd_ndisc_router *rt, const char **ret
         return 0;
 }
 
+static int ndisc_get_dns_name(const uint8_t *optval, size_t optlen, char **ret) {
+        _cleanup_free_ char *name = NULL;
+        int r;
+
+        assert(optval || optlen == 0);
+        assert(ret);
+
+        r = dns_name_from_wire_format(&optval, &optlen, &name);
+        if (r < 0)
+                return r;
+        if (r == 0 || optlen != 0)
+                return -EBADMSG;
+
+        *ret = TAKE_PTR(name);
+        return r;
+}
+
+int sd_ndisc_router_encrypted_dns_get_dnr(sd_ndisc_router *rt, ResolverData **ret) {
+        uint8_t *nd_opt_encrypted_dns = NULL;
+        size_t length;
+        int r;
+
+        assert_return(rt, -EINVAL);
+        assert_return(ret, -EINVAL);
+
+        r = sd_ndisc_router_option_is_type(rt, SD_NDISC_OPTION_ENCRYPTED_DNS);
+        if (r < 0)
+                return r;
+        if (r == 0)
+                return -EMEDIUMTYPE;
+
+        r = sd_ndisc_router_option_get_raw(rt, (void *)&nd_opt_encrypted_dns, &length);
+        if (r < 0)
+                return r;
+
+        assert(length % 8 == 0);
+        if (length == 0)
+                return -EBADMSG;
+
+        uint8_t *optval = &nd_opt_encrypted_dns[2];
+        size_t optlen = 8 * length - 2; /* units of octets */
+        size_t offset = 0;
+
+        _cleanup_(dnr_resolver_data_free_allp) ResolverData *res = new0(ResolverData, 1);
+        if (!res)
+                return -ENOMEM;
+        res->priority = unaligned_read_be16(&optval[offset]);
+        offset += sizeof(uint16_t);
+        offset += sizeof(uint32_t); /* Lifetime field. Accessed with *lifetime_timestamp functions. */
+
+        if (offset + sizeof(uint16_t) > optlen)
+                return -EBADMSG;
+
+        /* adn field (length + dns-name) */
+        size_t ilen = unaligned_read_be16(&optval[offset]);
+        offset += sizeof(uint16_t);
+        if (offset + ilen > optlen)
+                return -EBADMSG;
+
+        r = ndisc_get_dns_name(&optval[offset], ilen, &res->auth_name);
+        if (r < 0)
+                return r;
```

Comment:
Before I forget it seems `r` should be converted into `-EBADMSG` unless it's `-ENOMEM` by analogy with https://github.com/systemd/systemd/pull/31171/files. `dns_name_from_wire_format` can return `-EMSGSIZE` and some other errors coming from `dns_label_escape`.